### PR TITLE
Automated model interface generation on aws llms

### PR DESCRIPTION
**Pull request by b4sjoo**
### Description
This PR will enable automated model interface generation on AWS managed llms (Bedrock models), to unblock Fractal team
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLGuardrailsIT.java [hunk 1]
```diff
@@ -423,7 +427,11 @@ protected Response registerRemoteModelWithLocalRegexGuardrails(String name, Stri
             + "      ],\n"
             + "      \"regex\": [\"regex1\", \"regex2\"]\n"
             + "    }\n"
-            + "  }\n"
+            + "},\n"
```
Comment (mingshl) [inline plugin/src/test/java/org/opensearch/ml/rest/RestMLGuardrailsIT.java:430]: the IT only covers the general model general interface, please add the IT for aws llm interface later

**Review by Zhangxunmt — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/action/register/TransportRegisterModelAction.java [hunk 2]
```diff
@@ -239,7 +240,14 @@ private void doRegister(MLRegisterModelInput registerModelInput, ActionListener<
             if (Strings.isNotBlank(registerModelInput.getConnectorId())) {
                 connectorAccessControlHelper.validateConnectorAccess(client, registerModelInput.getConnectorId(), ActionListener.wrap(r -> {
                     if (Boolean.TRUE.equals(r)) {
-                        createModelGroup(registerModelInput, listener);
+                        if (registerModelInput.getModelInterface() == null) {
+                            mlModelManager.getConnector(registerModelInput.getConnectorId(), ActionListener.wrap(connector -> {
```
Comment (ylwu-amzn) [inline plugin/src/main/java/org/opensearch/ml/action/register/TransportRegisterModelAction.java:244]: Should we check if it's internal connector inside model ?

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLGuardrailsIT.java (hunk 1 above)
Comment (b4sjoo) [reply plugin/src/test/java/org/opensearch/ml/rest/RestMLGuardrailsIT.java:430]: This is to resolve IT failure, we have complete interface IT: https://github.com/opensearch-project/ml-commons/pull/2394

File: common/src/main/java/org/opensearch/ml/common/utils/ModelInterfaceUtils.java [hunk 3]
```diff
@@ -0,0 +1,623 @@
...
+import org.opensearch.ml.common.MLModel;
+import org.opensearch.ml.common.connector.Connector;
+import org.opensearch.ml.common.transport.register.MLRegisterModelInput;
+
+import java.util.Map;
+
+@Log4j2
+public class ModelInterfaceUtils {
+
+    private static final String GENERAL_CONVERSATIONAL_MODEL_INTERFACE_INPUT = "{\n" +
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/utils/ModelInterfaceUtils.java:18]: Should we maintain any kind of version for Model Interface? What happens if we want to add another field to the model interface later?

**Review by rbhavna — APPROVED** [review]

File: plugin/src/main/java/org/opensearch/ml/action/register/TransportRegisterModelAction.java [hunk 4]
```diff
@@ -239,7 +240,14 @@ private void doRegister(MLRegisterModelInput registerModelInput, ActionListener<
             if (Strings.isNotBlank(registerModelInput.getConnectorId())) {
                 connectorAccessControlHelper.validateConnectorAccess(client, registerModelInput.getConnectorId(), ActionListener.wrap(r -> {
                     if (Boolean.TRUE.equals(r)) {
-                        createModelGroup(registerModelInput, listener);
+                        if (registerModelInput.getModelInterface() == null) {
```
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/register/TransportRegisterModelAction.java:243]: are we expecting a new field in model index? I don't see any declaration of this new field? Or did you already do that in any other PR?

File: plugin/src/test/java/org/opensearch/ml/rest/RestMLGuardrailsIT.java [hunk 5]
```diff
@@ -366,7 +366,11 @@ protected Response registerRemoteModel(String modelGroupName, String name, Strin
             + "  \"description\": \"test model\",\n"
             + "  \"connector_id\": \""
             + connectorId
-            + "\"\n"
+            + "\",\n"
+            + "  \"interface\": {\n"
```
Comment (dhrubo-os) [inline plugin/src/test/java/org/opensearch/ml/rest/RestMLGuardrailsIT.java:370]: Is that a new field in the connector index?

Comment (b4sjoo) [reply plugin/src/test/java/org/opensearch/ml/rest/RestMLGuardrailsIT.java:370]: See above, it's in model index

File: common/src/main/java/org/opensearch/ml/common/utils/ModelInterfaceUtils.java (hunk 3 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/utils/ModelInterfaceUtils.java:18]: It's will be bounded by blueprints in the future. So if we want add another field we just directly introduce a new variable

File: plugin/src/main/java/org/opensearch/ml/action/register/TransportRegisterModelAction.java (hunk 2 above)
Comment (dhrubo-os) [inline plugin/src/main/java/org/opensearch/ml/action/register/TransportRegisterModelAction.java:244]: can't you stash thread context here before invoking this method?

Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/action/register/TransportRegisterModelAction.java:244]: Well yeah this is something we can try

**Pull request merged**